File: ai/project-mcp.mjs
// Project-only MCP configuration. Tested by project-mcp-test.sh.
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { isDeepStrictEqual } from 'node:util';
import { spawnSync } from 'node:child_process';

const here = path.dirname(fileURLToPath(import.meta.url));
const launcher = 'bucket=$(CDPATH= cd -P "$HOME/.kk-flavor" && pwd -P) || exit; exec "$bucket/../mcp-env.sh" "$@"';
const open = '# kk-flavor-mcp:begin';
const close = '# kk-flavor-mcp:end';

function parseArguments() {
  let agent, project;
  let isDryRun = false, isUninstall = false;
  for (const arg of process.argv.slice(2)) {
    if (arg === '--agent=claude' || arg === '--agent=codex') agent = arg.slice(8);
    else if (arg === '--dry-run') isDryRun = true;
    else if (arg === '--uninstall') isUninstall = true;
    else if (arg.startsWith('-') || project) throw new Error(`unknown or extra argument: ${arg}`);
    else project = arg;
  }
  if (!agent || !project) throw new Error('--agent=claude|codex and a project directory are required');
  project = fs.realpathSync(project);
  if (process.env.HOME && project === fs.realpathSync(process.env.HOME)) throw new Error('the home directory is not a project target');
  if (!fs.statSync(project).isDirectory()) throw new Error('project must be a directory');
  return {agent, project, isDryRun, isUninstall};
}

function readPublicServers() {
  const source = fs.readFileSync(path.join(here, 'mcp.jsonc'), 'utf8').replace(/^\s*\/\/.*$/gm, '');
  const servers = JSON.parse(source).mcpServers;
  if (!servers || Array.isArray(servers) || typeof servers !== 'object') throw new Error('public mcpServers must be an object');
  return Object.fromEntries(Object.entries(servers).map(([name, server]) => {
    if (!/^[a-zA-Z0-9_-]+$/.test(name) || server.type !== 'stdio' || server.command !== '@CONFIGS@/mcp-env.sh' ||
        !Array.isArray(server.args) || !server.args.every(arg => typeof arg === 'string' && !arg.includes('\0')) ||
        Object.keys(server).some(key => !['type', 'command', 'args'].includes(key))) {
      throw new Error(`public server ${name} needs an explicit portable project mapping`);
    }
    return [name, {type:'stdio', command:'sh', args:['-c', launcher, 'kk-flavor-mcp', ...server.args]}];
  }));
}

function readConfig(file) {
  try {
    const stat = fs.lstatSync(file);
    if (!stat.isFile() || stat.nlink !== 1) throw new Error(`${file} must be a regular unlinked file`);
    return fs.readFileSync(file, 'utf8');
  } catch (error) {
    if (error.code === 'ENOENT') return '';
    throw error;
  }
}

function mergeClaude({text, servers, isUninstall}) {
  const config = text ? JSON.parse(text) : {};
  if (!config || typeof config !== 'object' || Array.isArray(config)) throw new Error('.mcp.json must be an object');
  const current = config.mcpServers ?? {};
  if (!current || typeof current !== 'object' || Array.isArray(current)) throw new Error('mcpServers must be an object');
  let hasChanged = false;
  for (const [name, server] of Object.entries(servers)) {
    if (Object.hasOwn(current, name)) {
      if (!isDeepStrictEqual(current[name], server)) {
        if (isUninstall) continue;
        throw new Error(`project MCP server ${name} already exists with different settings`);
      }
      if (isUninstall) { delete current[name]; hasChanged = true; }
    } else if (!isUninstall) { current[name] = server; hasChanged = true; }
  }
  if (!hasChanged) return text;
  config.mcpServers = current;
  return JSON.stringify(config, undefined, 2) + '\n';
}

function mergeCodex({text, servers, isUninstall}) {
  const region = `${open}\n` + Object.entries(servers).map(([name, server]) =>
    `[mcp_servers.${name}]\ncommand = "sh"\nargs = ${JSON.stringify(server.args)}\n`).join('\n') + `${close}\n`;
  if (text.includes("'''") || text.includes('\"\"\"')) throw new Error('multiline TOML strings require an explicit MCP merge');
  let outside = text;
  const start = text.indexOf(open), end = text.indexOf(close);
  if (start >= 0 || end >= 0) {
    if (start < 0 || end < start || text.indexOf(open, start + open.length) >= 0 || text.indexOf(close, end + close.length) >= 0) {
      throw new Error('incomplete or repeated project MCP markers');
    }
    if ((start > 0 && text[start - 1] !== '\n') || (end > 0 && text[end - 1] !== '\n') ||
        !['', '\n'].includes(text.slice(end + close.length, end + close.length + 1))) {
      throw new Error('project MCP markers must occupy whole lines');
    }
    const owned = text.slice(start, end + close.length) + '\n';
    if (owned !== region) throw new Error('project MCP region was edited; preserve or remove it explicitly');
    const suffix = text.slice(end + close.length);
    const firstFollowingSetting = suffix.split('\n').find(line => line.trim() && !/^\s*#/.test(line));
    if (firstFollowingSetting && !/^\s*\[/.test(firstFollowingSetting)) {
      throw new Error('settings after the MCP region belong to a managed server; move them before removing it');
    }
    outside = text.slice(0, start) + text.slice(end + close.length).replace(/^\n/, '');
  }
  if (isUninstall) return outside;
  // Recognize ordinary server tables only. Other spellings could redefine the parent table or a
  // managed server, so leave those files for an explicit edit instead of guessing at TOML syntax.
  for (const line of outside.split('\n')) {
    if (/^\s*#/.test(line)) continue;
    if (/\\[uU]/.test(line)) throw new Error('escaped TOML requires an explicit MCP merge');
    if (!line.includes('mcp_servers')) continue;
    const match = line.match(/^\s*\[mcp_servers\.([a-zA-Z0-9_-]+)(?:\.[a-zA-Z0-9_-]+)*\]\s*(?:#.*)?$/);
    if (!match || Object.hasOwn(servers, match[1])) throw new Error('existing MCP TOML conflicts or uses ambiguous syntax; merge it explicitly');
  }
  if (start >= 0) return text;
  return outside + (outside && !outside.endsWith('\n') ? '\n' : '') + region;
}

function writeConfig({file, text, previous}) {
  if (text === previous) return;
  fs.mkdirSync(path.dirname(file), {recursive:true});
  const temporary = `${file}.kk-mcp-${process.pid}`;
  let wasCreated = false;
  try {
    const mode = fs.existsSync(file) ? fs.statSync(file).mode : 0o644;
    fs.writeFileSync(temporary, text, {flag:'wx', mode});
    wasCreated = true;
    fs.renameSync(temporary, file);
  } finally {
    if (wasCreated && fs.existsSync(temporary)) fs.unlinkSync(temporary);
  }
}

try {
  const options = parseArguments();
  const directory = path.join(options.project, '.codex');
  if (options.agent === 'codex') {
    try {
      if (!fs.lstatSync(directory).isDirectory()) throw new Error('.codex must be a real directory, not a link');
    } catch (error) {
      if (error.code !== 'ENOENT') throw error;
    }
  }
  const file = path.join(options.project, options.agent === 'claude' ? '.mcp.json' : '.codex/config.toml');
  if (!options.isUninstall) {
    const relative = path.relative(options.project, file);
    const git = (...args) => spawnSync('git', ['-C', options.project, ...args], {encoding:'utf8'});
    if (git('rev-parse', '--is-inside-work-tree').status === 0) {
      const ignored = git('check-ignore', '--quiet', '--', relative);
      if (ignored.status === 0) {
        throw new Error(`${relative} is ignored by Git; adjust the project ignore rules to include this configuration, then rerun and review it before committing`);
      }
      if (ignored.status !== 1) throw new Error(`cannot check whether Git will track ${relative}`);
    }
  }
  const previous = readConfig(file);
  const input = {...options, text:previous, servers:readPublicServers()};
  const text = options.agent === 'claude' ? mergeClaude(input) : mergeCodex(input);
  if (!options.isDryRun) writeConfig({file, text, previous});
  console.log(`project MCP: ${options.isDryRun ? 'would update' : 'checked'} ${file}`);
} catch (error) {
  console.error(`project MCP: ${error.message}`);
  process.exitCode = 1;
}

File: ai/tools/project-mcp/projectmcp.go
//	usage: project-mcp.sh --agent=claude|codex [--dry-run] [--uninstall] <project>
//
// The servers come from `ai/mcp.jsonc`, the committed half of the declaration. `mcp.private.jsonc`
// stays closed here. A project file is committed and reviewed, so anything carrying a credential or
// naming an internal host belongs in the user-scope sync alone.

// What lands in the project differs from what the user scope registers. A user-scope entry names this
// checkout by absolute path. A project file is shared with everyone who clones the project, so each
// server is rewritten to reach the launcher through `$HOME/.kk-flavor`, the path every install of
// this flavor has.

// Configure the public MCP servers in one project's own client files, and never in user settings.
// `usage` in this file states the grammar the stub takes, and the command's own header repeats it.
package projectmcp

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"

	"configs/ai/tools/repo"
	"configs/ai/tools/shell"
)

// Exit codes on the tools' shared vocabulary. 2 is a grammar this tool failed to parse, and it
// attempted no write. 1 is a project or a config file it refuses to write.
const (
	exitDone     = 0
	exitRefused  = 1
	exitBadUsage = 2
)

const claudeAgent = "claude"

const codexAgent = "codex"

var agents = []string{claudeAgent, codexAgent}

// Where each client keeps a project's own server list.
var configFiles = map[string]string{
	claudeAgent: ".mcp.json",
	codexAgent:  ".codex/config.toml",
}

// Every message a human reads from this tool, whichever client it was asked about.
const label = "project MCP"

// `configsDir` is the directory holding `mcp.jsonc`, which is the tool's own directory. The caller
// resolves it from the path the stub was invoked by.

// `home` is the home directory refuseHome compares against. It arrives as a value, so no case in the
// suite can reach the owner's own by reading the environment.

// Run executes one invocation and returns its exit code.
func Run(self string, args []string, configsDir, home string, git repo.Git, stdout, stderr io.Writer) int {
	run, code, hasStopped := parseArguments(self, args, stdout, stderr)
	if hasStopped {
		return code
	}
	run.configsDir, run.home, run.git = configsDir, home, git
	if err := run.do(stdout); err != nil {
		fmt.Fprintf(stderr, "%s: %s\n", label, err)
		return exitRefused
	}
	return exitDone
}

// One invocation's fixed context. Every step needs the project and the agent. A run that read one
// project while writing another's file is the inconsistency these two fields cannot express.
type invocation struct {
	agent       string
	project     string
	isDryRun    bool
	isUninstall bool

	configsDir string
	home       string
	git        repo.Git
}

// The third value says whether the run stops here. A refused invocation and a printed help both do,
// and they exit differently. A caller holding only the code reads exit 0 from the help arm as "parsed
// fine, carry on". That is how an empty invocation reaches the filesystem.
func parseArguments(self string, args []string, stdout, stderr io.Writer) (invocation, int, bool) {
	run := invocation{}
	for _, arg := range args {
		switch {
		case arg == "-h" || arg == "--help":
			fmt.Fprintln(stdout, usage(self))
			return run, exitDone, true
		case strings.HasPrefix(arg, "--agent=") && isKnownAgent(strings.TrimPrefix(arg, "--agent=")):
			run.agent = strings.TrimPrefix(arg, "--agent=")
		case arg == "--dry-run":
			run.isDryRun = true
		case arg == "--uninstall":
			run.isUninstall = true
		case strings.HasPrefix(arg, "-"):
			return run, badUsage(self, stderr, "unknown option %s", arg), true
		case run.project != "":
			return run, badUsage(self, stderr, "select one project"), true
		default:
			run.project = arg
		}
	}
	if run.agent == "" || run.project == "" {
		return run, badUsage(self, stderr,
			"--agent=%s and an existing project directory are required", selector()), true
	}
	if info, err := os.Stat(run.project); err != nil || !info.IsDir() {
		return run, badUsage(self, stderr,
			"--agent=%s and an existing project directory are required", selector()), true
	}
	return run, exitDone, false
}

// A refusal goes to stderr, because a caller capturing stdout is reading the line naming the file
// this tool checked. The help arm is the exception and prints where a human asked for it.
func badUsage(self string, stderr io.Writer, format string, args ...any) int {
	fmt.Fprintf(stderr, "%s: %s\n", label, fmt.Sprintf(format, args...))
	fmt.Fprintln(stderr, usage(self))
	return exitBadUsage
}

func usage(self string) string {
	return fmt.Sprintf("usage: %s --agent=%s [--dry-run] [--uninstall] <project>", self, selector())
}

func selector() string {
	return strings.Join(agents, "|")
}

func isKnownAgent(agent string) bool {
	for _, known := range agents {
		if agent == known {
			return true
		}
	}
	return false
}

// The whole of one run: resolve the project, work out what the file should hold, and write it.
func (run *invocation) do(stdout io.Writer) error {
	// The path is made absolute as well as symlink-free, because refuseHome compares it against paths
	// that already are. `.` is what a human standing in their home types, and a relative spelling
	// reaching the home guard matches none of what it guards.
	project, err := shell.RealPath(run.project)
	if err != nil {
		return err
	}
	run.project = project
	if err = run.refuseHome(); err != nil {
		return err
	}
	if err = run.refuseCodexDirectoryLink(); err != nil {
		return err
	}

	file := filepath.Join(run.project, filepath.FromSlash(configFiles[run.agent]))
	if !run.isUninstall {
		if err = run.refuseUntrackedConfig(file); err != nil {
			return err
		}
	}
	servers, err := readPublicServers(run.configsDir)
	if err != nil {
		return err
	}
	previous, err := readConfig(file)
	if err != nil {
		return err
	}
	text, err := run.merge(previous, servers)
	if err != nil {
		return err
	}
	if !run.isDryRun {
		if err = writeConfig(file, text, previous); err != nil {
			return err
		}
	}
	verb := "checked"
	if run.isDryRun {
		verb = "would update"
	}
	fmt.Fprintf(stdout, "%s: %s %s\n", label, verb, file)
	return nil
}

func (run *invocation) merge(previous string, servers []projectServer) (string, error) {
	if run.agent == claudeAgent {
		return mergeClaude(previous, servers, run.isUninstall)
	}
	return mergeCodex(previous, servers, run.isUninstall)
}

// The home directory is where the user-scope sync writes, and this tool writes project files. Given
// one as the project it would put a project's server list into the human's home, where every client
// reads it as a project it is standing in.
func (run *invocation) refuseHome() error {
	if run.home == "" {
		return nil
	}
	home, err := shell.RealPath(run.home)
	if err != nil {
		return nil
	}
	if run.project == home {
		return fmt.Errorf("the home directory is not a project target")
	}
	return nil
}

// Codex's config sits one directory down, so this tool may have to create `.codex/`. A link there
// points that write outside the project, at another project's configuration or at whatever else the
// link names. A `.codex` that is not a real directory stops the run.
func (run *invocation) refuseCodexDirectoryLink() error {
	if run.agent != codexAgent {
		return nil
	}
	directory := filepath.Join(run.project, ".codex")
	info, err := os.Lstat(directory)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return err
	}
	if !info.IsDir() {
		return fmt.Errorf(".codex must be a real directory, not a link")
	}
	return nil
}

// The ignore rules are the project's, so this reports them and leaves them alone. An uninstall skips
// the check, because removing servers from a file the project ignores is still worth doing.

// A project's MCP configuration is meant to be committed and reviewed, which is the whole difference
// between it and the user-scope registry. A config written into a path the project ignores reaches no
// other clone, and it differs silently from what every clone has.
func (run *invocation) refuseUntrackedConfig(file string) error {
	if _, err := run.git.TopLevel(run.project); err != nil {
		return nil
	}
	relative, err := filepath.Rel(run.project, file)
	if err != nil {
		return err
	}
	relative = filepath.ToSlash(relative)
	ignored, err := run.git.Ignored(run.project, []string{relative})
	if err != nil {
		return fmt.Errorf("cannot check whether Git will track %s", relative)
	}
	if ignored[relative] {
		return fmt.Errorf("%s is ignored by Git; adjust the project ignore rules to include this "+
			"configuration, then rerun and review it before committing", relative)
	}
	return nil
}
